---
title: "assignment week 6"
author: "Charles Cheruparambil"
date: "2023-04-20"
output: html_document
---
```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("memes")
BiocManager::install("motifmatchr")
BiocManager::install("MotifDb")
BiocManager::install("PWMEnrich")
BiocManager::install("rGADEM")
suppressPackageStartupMessages({
  library(AnnotationHub)
  library(GenomicRanges)
  library(ggplot2)
  library(memes) # for the meme-based methods -- COMMENT OUT when using alternatives
  library(motifmatchr)
  library(Biostrings) # for handling sequences
  library(MotifDb) # database of motifs
  library(TFBSTools) # for handling some motif formats
  library(universalmotif) # for converting motifs to various formats
  library(PWMEnrich) # for R-based motif enrichment analysis
})
```
#Choose a transcription factor, e.g. CREB1, REST, GATA5, EGR1, GCR (or any of your choice that has a motif and available ChIPseq data):
Target: REST
Assay: ChIP-seq (TF ChIP-seq)
Biosample summary: Mus musculus myocyte originated from C2C12
Biosample Type: in vitro differentiated cells


#Download the (e.g. Mouse) peaks for that factor (in whatever cell type)
# Obtaining peak sequences
```{r, REST}
download.file("https://www.encodeproject.org/files/ENCFF368VWJ/@@download/ENCFF368VWJ.bed.gz", dest="REST_TF_ChIPseq.bed.gz")
peaks <- rtracklayer::import("REST_TF_ChIPseq.bed.gz", format="NarrowPeak")
seqlevelsStyle(peaks) <- "Ensembl"  # to change the convention of the chromosome names to ensembl (i.e. without 'chr')
peaks_REST_chr1 <- peaks[seqnames(peaks)=="1"]
```

Since the motif of a transcription factor is typically found towards the center or summit of the peak, it is common practice to look for motifs around the center of peaks. 
How far around depends on the resolution of your data, i.e. how narrow are your peaks (can be between 50bp to 500bp around), as well as the nature of the signal from which the peaks were obtained.
The first thing we do, then, is to extract the regions around the peaks, and then obtain the sequence from those regions:

```{r}
peak_centers <- resize(peaks_REST_chr1, fix="center", width=100)
# we get the genome sequence:
ah <- AnnotationHub(localHub=FALSE)
genome <- ah[["AH68356"]]
# we get the sequences corresponding to the peak centers:
peak_seqs <- memes::get_sequence(peak_centers, genome)
# peak_seqs <- Biostrings::getSeq(genome, peak_centers)
# names(peak_seqs) <- as.character(granges(peak_centers))
peak_seqs
as.character(peak_seqs[1])
```

#Identify the instances of the factor's motif
# Motif scanning

Motif **scanning** aims at finding the _occurrences_ of a known motif in a set of sequences.
As this is methodologically fairly simple, it doesn't really matter what method one uses, though results may differ because of how the thresholds are set.

## Getting the desired motif

```{r}
# we search for "REST" in the motif database
motifs <- query(MotifDb, "REST")
# there are several matching motifs:
names(motifs)
```


```{r}
# we select one:
motif <- motifs[["Mmusculus-HOCOMOCOv10-REST_MOUSE.H10MO.A"]]
# we visualize it:
view_motifs(motif)
```

## Scanning a set of sequences using motifmatchr (doesn't require Meme installation):

```{r}
# if you don't already have the genome in fasta format saved somewhere, convert it to that format:
Biostrings::writeXStringSet(rtracklayer::import(genome), "genome.fa")
# we also need to convert the motif to a format that this package will accept
motif2 <- convert_motifs(motif, class="TFBSTools-PFMatrix")
# if we had a list of motifs instead, we'd convert them like this:
# motifs2 <- setNames(do.call(PFMatrixList, convert_motifs(motifs, class="TFBSTools-PFMatrix")), names(motifs))
moi2 <- motifmatchr::matchMotifs(motif2, subject=peak_centers, genome=Rsamtools::FaFile("genome.fa"), out="positions")[[1]]
head(moi2)
```

### Another R alternative...

For those users who can't install memes & where `motifmatchr::matchMotifs` somehow doesn't work...

```{r}
#' findMotifInstances
#'
#' A wrapper around TFBSTools for scanning motif occurence, handling the 
#' coordinate conversion as `memes` does.
#' 
#' @param seqs A set of sequences, e.g. `DNAStringSet`, optionally with 
#'   coordinate as names to enable conversion. (e.g. use 
#'   `names(seqs) <- as.character(granges(peaks))` to set them)
#' @param motif A motif, in any format recognized by `universalmotif`
#' @param keepMatchedSeq Logical; whether to keep the matched sequence.
#' @param ... Passed to `TFBSTools::searchSeq`; can for instance be used to set
#'   the number of threads to use, e.g. with `mc.cores=2`
#'
#' @return A `GRanges` object
findMotifInstances <- function(seqs, motif, keepMatchedSeq=FALSE, ...){
  motif <- universalmotif::convert_motifs(motif, "TFBSTools-PWMatrix")
  x <- suppressWarnings(TFBSTools::searchSeq(motif, subject=seqs))
  peaks <- strsplit(gsub("-",":",names(seqs)), ":")
  if(!is.null(names(seqs)) && all(lengths(peaks)==3)){ # convert relative coordinates to absolute
    chrs <- sapply(peaks,FUN=function(x) x[1])
    offsets <- sapply(peaks,FUN=function(x) as.integer(x[2]))-1L
    i <- rep(seq_along(x),lengths(x))
    return(GRanges(chrs[i], 
                   IRanges(offsets[i]+as.integer(unlist(lapply(x, start))),
                           offsets[i]+as.integer(unlist(lapply(x, end)))),
                   strand=unlist(lapply(x, strand)),
                   score=as.numeric(unlist(lapply(x, FUN=function(x) score(x))))
                   ))
  }
  x <- as(x, "GRanges")
  score(x) <- x$absScore
  keepFields <- c("score","relScore")
  if(keepMatchedSeq) keepFields <- c(keepFields, "siteSeqs")
  mcols(x) <- mcols(x)[,intersect(colnames(mcols(x)), keepFields)]
  x
}
```

```{r}
moi <- findMotifInstances(peak_seqs, motif)
```

```{r}
library(rGADEM)
moi <- findMotifInstances(peak_seqs,motif)
```


## Comparison of the results

```{r, eval=FALSE}
epiwraps::regionUpset(list(fimo=moi, motifmatchr=moi2))
boxplot(moi$score~overlapsAny(moi, moi2))
```


```{r}
# peaks with motif:
peaks_w_motif <- peaks_REST_chr1[overlapsAny(peaks_REST_chr1,moi)]
length(peaks_w_motif)
length(peaks_REST_chr1)
length(peaks_w_motif)/length(peaks_REST_chr1)
```

#Answer the following questions:
#Of all the peaks, what proportion contains a motif for the factor?
#Expected form of an answer: of the XX peaks, XX (XX%) contain a motif
```{r}
# peaks with motif:
peaks_w_motif <- peaks_REST_chr1[overlapsAny(peaks_REST_chr1,moi)]
# calculate the proportion of peaks with the motif
prop_peaks_w_motif <- sprintf("%.2f%%", length(peaks_w_motif) / length(peaks_REST_chr1) * 100)
# print the answer
cat(paste0("Of all the peaks, ", length(peaks_w_motif), " (", prop_peaks_w_motif, ") contain a motif."))
```

#Of all instances of that motif in the genome (or in one chromosome), what proportion is bound by the factor (i.e. has a peak)?
#Expected form of an answer: of the XX motif instances, XX (XX%) overlap a peak

```{r}
# calculate the total number of motif instances in the chromosome
total_motif_instances <- length(moi)
# calculate the number of motif instances that overlap a peak
motif_instances_with_peak <- length(unique(moi[overlapsAny(moi, peaks_REST_chr1)]))
# calculate the proportion of motif instances with a peak
prop_motif_instances_with_peak <- motif_instances_with_peak / total_motif_instances
# print the results
cat("Of all the", total_motif_instances, "motif instances in the chromosome,", motif_instances_with_peak, "(", round(prop_motif_instances_with_peak*100, 2), "%)", "overlap a peak.")
```




